### PR TITLE
Imu

### PR DESCRIPTION
**Pull request by kam3k**
Merged (and resolved merge conflicts) the IMU plugin written by marbosjo.

### CODE REVIEW

File: flatland_plugins/include/flatland_plugins/imu.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
+/*
+ *  ______                   __  __              __
+ * /\  _  \           __    /\ \/\ \            /\ \__
+ * \ \ \L\ \  __  __ /\_\   \_\ \ \ \____    ___\ \ ,_\   ____
+ *  \ \  __ \/\ \/\ \\/\ \  /'_` \ \ '__`\  / __`\ \ \/  /',__\
+ *   \ \ \/\ \ \ \_/ |\ \ \/\ \L\ \ \ \L\ \/\ \L\ \ \ \_/\__, `\
+ *    \ \_\ \_\ \___/  \ \_\ \___,_\ \_,__/\ \____/\ \__\/\____/
+ *     \/_/\/_/\/__/    \/_/\/__,_ /\/___/  \/___/  \/__/\/___/
+ * @copyright Copyright 2017 Avidbots Corp.
+ * @name	diff_drive.h
```
Comment (avidbots-kenneth) [inline flatland_plugins/include/flatland_plugins/imu.h:10]: Should be `imu.h` and "IMU plugin" in the brief.
Also, should we be using a newer set of dates and authors here?

File: flatland_plugins/test/imu_test.test [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<!--
+Test launchfile for model_tf_publisher_test
```
Comment (avidbots-kenneth) [inline flatland_plugins/test/imu_test.test:2]: Change to imu_test

File: flatland_plugins/include/flatland_plugins/imu.h (hunk 1 above)
Comment (kam3k) [reply flatland_plugins/include/flatland_plugins/imu.h:10]: I'll update this to `imu.h` and update the brief. But yeah as for the other stuff, I'll do whatever you'd like. My role here was taking an old PR and bringing it up to date with a bunch of merge conflict resolution. But I'm not the original author.

Comment (avidbots-kenneth) [reply flatland_plugins/include/flatland_plugins/imu.h:10]: Okay we can leave the date and author as-is then.

File: flatland_plugins/include/flatland_plugins/imu.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  b2Vec2 linear_vel_local_prev;
+  double pub_rate_;
+  /**
+   * @name          OnInitialize
+   * @brief         override the BeforePhysicsStep method
+   * @param[in]     config The plugin YAML node
+   */
+  void OnInitialize(const YAML::Node& config) override;
+  /**
+   * @name          BeforePhysicsStep
```
Comment (avidbots-kenneth) [inline flatland_plugins/include/flatland_plugins/imu.h:88]: `AfterPhysicsStep`

File: flatland_plugins/include/flatland_plugins/imu.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  ros::Publisher ground_truth_pub_;
+  Body* body_;
+  sensor_msgs::Imu imu_msg_;
+  sensor_msgs::Imu ground_truth_msg_;
+  UpdateTimer update_timer_;
+  bool enable_imu_pub_;  ///< YAML parameter to enable odom publishing
+
+  std::default_random_engine rng_;
+  std::array<std::normal_distribution<double>, 9> noise_gen_;
+  geometry_msgs::TransformStamped imu_tf_;   ///< tf from body to laser frame
```
Comment (avidbots-kenneth) [inline flatland_plugins/include/flatland_plugins/imu.h:75]: to IMU frame, + next line as well

File: flatland_plugins/include/flatland_plugins/imu.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  /**
+   * @name          OnInitialize
+   * @brief         override the BeforePhysicsStep method
+   * @param[in]     config The plugin YAML node
+   */
+  void OnInitialize(const YAML::Node& config) override;
+  /**
+   * @name          BeforePhysicsStep
+   * @brief         override the BeforePhysicsStep method
+   * @param[in]     config The plugin YAML node
```
Comment (avidbots-kenneth) [inline flatland_plugins/include/flatland_plugins/imu.h:90]: timekeeper

**Comment by avidbots-kenneth** [comment]
We should add a tutorial like the other plugins in https://github.com/avidbots/flatland/tree/master/docs/included_plugins which gets posted in here https://flatland-simulator.readthedocs.io/en/latest/included_plugins/diff_drive.html for example.

File: flatland_plugins/include/flatland_plugins/imu.h (hunk 4 above)
Comment (kam3k) [reply flatland_plugins/include/flatland_plugins/imu.h]: Fixed, thanks.

File: flatland_plugins/include/flatland_plugins/imu.h (hunk 3 above)
Comment (kam3k) [reply flatland_plugins/include/flatland_plugins/imu.h]: Fixed, thanks.

File: flatland_plugins/test/imu_test.test (hunk 2 above)
Comment (kam3k) [reply flatland_plugins/test/imu_test.test]: Fixed, thanks.

**Review by josephduchesne — APPROVED** [review]

**Comment by kam3k** [comment]
> We should add a tutorial like the other plugins in https://github.com/avidbots/flatland/tree/master/docs/included_plugins which gets posted in here https://flatland-simulator.readthedocs.io/en/latest/included_plugins/diff_drive.html for example.

I've just written and added a tutorial for this plugin.

**Review by avidbots-kenneth — APPROVED** [review]

**Pull request merged**